multicluster/cmd/gateways.go

```diff
@@ -1,95 +1,218 @@
 )
 
 type (
 	gatewaysOptions struct {
-		gatewayNamespace string
+		clusterName string
+		wait        time.Duration
+	}
+
+	gatewayMetrics struct {
+		clusterName string
+		metrics     []byte
+		err         error
+	}
+
+	gatewayStatus struct {
 		clusterName      string
-		timeWindow       string
+		alive            bool
+		numberOfServices int
+		latency          uint64
 	}
 )
 
+func newGatewaysOptions() *gatewaysOptions {
+	return &gatewaysOptions{
+		wait: 30 * time.Second,
+	}
+}
+
 func newGatewaysCommand() *cobra.Command {
 
-	opts := gatewaysOptions{}
+	opts := newGatewaysOptions()
 
 	cmd := &cobra.Command{
 		Use:   "gateways",
 		Short: "Display stats information about the gateways in target clusters",
 		Args:  cobra.NoArgs,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			req := &pb.GatewaysRequest{
-				RemoteClusterName: opts.clusterName,
-				GatewayNamespace:  opts.gatewayNamespace,
-				TimeWindow:        opts.timeWindow,
-			}
-
 			k8sAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext, impersonate, impersonateGroup, 0)
 			if err != nil {
 				return err
 			}
 
-			ctx := cmd.Context()
-
-			vizNs, err := k8sAPI.GetNamespaceWithExtensionLabel(ctx, vizCmd.ExtensionName)
+			// Get all the service mirror components in the
+			// linkerd-multicluster namespace which we'll collect gateway
+			// metrics from.
+			multiclusterNs, err := k8sAPI.GetNamespaceWithExtensionLabel(cmd.Context(), MulticlusterExtensionName)
 			if err != nil {
-				return fmt.Errorf("make sure the linkerd-viz extension is installed, using 'linkerd viz install' (%w)", err)
+				return fmt.Errorf("make sure the linkerd-multicluster extension is installed, using 'linkerd multicluster install' (%w)", err)
 			}
-
-			client, err := client.NewExternalClient(ctx, vizNs.Name, k8sAPI)
+			selector := fmt.Sprintf("component=%s", "linkerd-service-mirror")
+			pods, err := k8sAPI.CoreV1().Pods(multiclusterNs.Name).List(cmd.Context(), metav1.ListOptions{LabelSelector: selector})
 			if err != nil {
-				return err
-			}
-
-			resp, err := requestGatewaysFromAPI(client, req)
-			if err != nil {
-				fmt.Fprintln(os.Stderr, err)
+				fmt.Fprintf(os.Stderr, "failed to list pods in namespace %s: %s", multiclusterNs.Name, err)
 				os.Exit(1)
 			}
 
-			renderGateways(resp.GetOk().GatewaysTable.Rows, stdout)
+			var statuses []gatewayStatus
+			gatewayMetrics := getGatewayMetrics(k8sAPI, pods.Items, opts.wait)
+			for _, gateway := range gatewayMetrics {
+				if gateway.err != nil {
+					fmt.Fprintf(os.Stderr, "Failed to get gateway status for %s: %s\n", gateway.clusterName, gateway.err)
+					continue
+				}
+				if opts.clusterName != "" && gateway.clusterName != opts.clusterName {
+					continue
+				}
+				gatewayStatus := gatewayStatus{
+					clusterName: gateway.clusterName,
+				}
+
+				// Parse the gateway metrics so that we can extract liveness
+				// and latency information.
+				var metricsParser expfmt.TextParser
```

Comment:
is `TextToMetricFamilies` a static function in the package? It looks like we're aliasing the type here so it doesn't look like `expfmt.TextParser.TextToMetricFamilies`. We could probably also alias the import here if you think it'd be more explicit

```go
import (
  metricsParser "github.com/prometheus/common/expfmt"
)

...

parsedMetrics, err := metricsParser.TextParser.TextToMetricFamilies(..)
```

Reply:
Nope the issue here is that `TextToMetricFamilies` is a pointer method so we need an addressable `TextParser`.

We could change it to one line with `(&expfmt.TextParser{})` but that seems more confusing to read.

Reply:
Ahhhhh, well that makes sense. Yeah it's fine as it is.